### PR TITLE
Fix low prio rez pull

### PR DESCRIPTION
**Pull request by roberthartmann**
sorry deleted the old branch by mistake

### CODE REVIEW

File: luaui/Widgets/gui_top_bar.lua [hunk 1]
```diff
@@ -11,6 +11,44 @@ function widget:GetInfo()
 	}
 end
 
+-- for bp bar only
+local metalCostForCommander = 1250
```
Comment (sprunk) [inline luaui/Widgets/gui_top_bar.lua:15]: could take from the unit def

File: luaui/Widgets/gui_top_bar.lua [hunk 2]
```diff
@@ -11,6 +11,44 @@ function widget:GetInfo()
...
+
+
+local spSetUnitBuildSpeed = Spring.SetUnitBuildSpeed
+local spGetUnitIsBuilding = Spring.GetUnitIsBuilding
+local spGetUnitDefID = Spring.GetUnitDefID
+
+local cost={}
+local totalStallingM = 0
+local totalStallingE = 0
+local addStalling = 1
```
Comment (sprunk) [inline luaui/Widgets/gui_top_bar.lua:46]: this seems to only be 0 or 1, shouldn't be true/false if so?

File: luaui/Widgets/gui_top_bar.lua [hunk 3]
```diff
@@ -11,6 +11,44 @@ function widget:GetInfo()
...
+local unitBuildSpeed = {}
+
+local CMD_PRIORITY = 34571 --low prio builders
```
Comment (sprunk) [inline luaui/Widgets/gui_top_bar.lua:49]: doesn't BAR have a master list of magic constants?

File: luaui/Widgets/gui_top_bar.lua [hunk 4]
```diff
@@ -773,8 +862,12 @@ local function updateResbar(res)
 		local texSize = math_floor(iconSize * 2)
 		if res == 'metal' then
 			glTexture(":lr" .. texSize .. "," .. texSize .. ":LuaUI/Images/metal.png")
-		else
-			glTexture(":lr" .. texSize .. "," .. texSize .. ":LuaUI/Images/energy.png")
+		elseif res == 'energy' then
+
+			glTexture(":lr" .. texSize .. "," .. texSize .. ":LuaUI/Images/energy.png") 
+        else
+            glColor(0.8, 1, 0.8, 1)
```
Comment (sprunk) [inline luaui/Widgets/gui_top_bar.lua:869]: check if this is in the correct place (whitespace suggests a mistake, perhaps just fix whitespace)

File: luaui/Widgets/gui_top_bar.lua [hunk 5]
```diff
@@ -1343,6 +1619,35 @@ local function drawResBars()
...
+				glColor(1, 0, 0, 0.13 * allyteamOverflowingEnergy * blinkProgress)
+			elseif overflowingEnergy then
+				glColor(1, 1, 0, 0.05 * overflowingEnergy * (0.6 + (blinkProgress * 0.4)))
+			end
+			if allyteamOverflowingEnergy or overflowingEnergy then
+				glCallList(dlistResbar[res][4])
+			end
+			-- low energy background
+			if r[res][1] < 2000 then
+				local process = (r[res][1] / r[res][2]) * 13
```
Comment (sprunk) [inline luaui/Widgets/gui_top_bar.lua:1637]: what's the significance of the 13? could use a comment or a named constant

File: luaui/Widgets/gui_top_bar.lua [hunk 6]
```diff
@@ -2056,3 +2406,53 @@ function widget:SetConfigData(data)
...
 	end
 end
+
+
+function findBPCommand(unitID, ...) -- for bp bar only most likely
+    local commands = Spring.GetUnitCommands(unitID, -1)
+    local cmdList = {...} 
+    for i = 1, #commands do
+        for _, CMD in ipairs(cmdList) do
+            if commands[i].id == CMD or commands[i].id < 0 then
```
Comment (sprunk) [inline luaui/Widgets/gui_top_bar.lua:2416]: `CMD` is a table so the numerical id will never be equal to it, should be `CMD.SOMETHING`

File: luaui/Widgets/gui_top_bar.lua [hunk 7]
```diff
@@ -1176,11 +1329,132 @@ function widget:Update(dt)
...
+            end
+        end
+
+        totalReservedBPPercentage = 0
+        if totalBP > 0 then
+            totalReservedBPPercentage = (totalReservedBP / totalBP) * 100
+        end
+
+        table.insert(totalReservedBPData, totalReservedBPPercentage)
+        if #totalReservedBPData > 30 then
```
[the remark this answers is not in the document]
Comment (roberthartmann) [reply luaui/Widgets/gui_top_bar.lua:1405]: 10 time steps of 15 game frames

File: luaui/Widgets/gui_top_bar.lua [hunk 8]
```diff
@@ -801,7 +894,7 @@ local function updateResbar(res)
 
 	dlistResbar[res][2] = glCreateList(function()
 		-- Metalmaker Conversion slider
-		if res == 'energy' then
+		if res == 'energy' then
```
Comment (MasterBel2) [inline luaui/Widgets/gui_top_bar.lua:897]: Spaces

File: luaui/Widgets/gui_top_bar.lua [hunk 9]
```diff
@@ -145,21 +202,24 @@ local windRotation = 0
 
 local lastFrame = -1
 local topbarArea = {}
-local resbarArea = { metal = {}, energy = {} }
-local resbarDrawinfo = { metal = {}, energy = {} }
-local shareIndicatorArea = { metal = {}, energy = {} }
-local dlistResbar = { metal = {}, energy = {} }
+local resbarArea = { metal = {}, energy = {}, BP = {} }
+local resbarDrawinfo = { metal = {}, energy = {}, BP = {} }
+local shareIndicatorArea = { metal = {}, energy = {}, BP = {} } --get rid of BP
```
Comment (MasterBel2) [inline luaui/Widgets/gui_top_bar.lua:207]: This comment is confusing

File: luaui/Widgets/gui_top_bar.lua [hunk 10]
```diff
@@ -27,7 +65,7 @@ local height = orgHeight * (1 + (ui_scale - 1) / 1.7)
 
 local escapeKeyPressesQuit = false
 
-local relXpos = 0.3
+local relXpos = 0.25 -- ooo 0.3
```
Comment (MasterBel2) [inline luaui/Widgets/gui_top_bar.lua:68]: Unhelpful comment; either elaborate or remove?

**Review by MasterBel2 — CHANGES_REQUESTED** [review]
Looking pretty good, most comments are pretty minor things with a bunch of repeated concerns. Should hopefully be quick and easy to clean up

File: luaui/Widgets/gui_top_bar.lua [hunk 11]
```diff
@@ -2056,3 +2406,53 @@ function widget:SetConfigData(data)
...
+function InitUnit(unitID, unitDefID, unitTeam) -- needed for exact calculations
+    local unitDef = UnitDefs[unitDefID]
+    if (myTeamID == unitTeam) then
+        if unitDef and unitDef.buildSpeed and unitDef.buildSpeed > 0 and unitDef.canAssist then
+            local isFactory = 0 -- is factory is needed for bp bar only (switch metal cost on/off)
+            if unitDef.isFactory == true then
+                isFactory = 1
+            end
+
+            if isFactory == 0 then
```
[the remark this answers is not in the document]
Comment (thehobojoe) [reply luaui/Widgets/gui_top_bar.lua:2449]: I think being more verbose in this case is probably good, it's subjective but I find the current pattern more readable, I would just do 
`local isFactory = unitDef.isFactory or false` to keep it neater

**Comment by thehobojoe** [comment]
Since you've already received very detailed code review I'll just leave a higher level comment. While this new bar is very neat, not everyone is going to want it, and it needs to be configurable, there are two ways this could be achieved:

1. Simply put in a toggle. This can start with just a boolean in the file while you put it together, making sure everything is as it was before with this toggle set to false, once that's working you would need to add a flag to the settings somewhere to enable it. I know you're new to programming so this probably feels daunting, but don't be afraid to ask questions during the process. This has downsides however, not only will this widget simply have more complex code and be harder to parse, but it will also (marginally, granted) add performance overhead.
2. Make it a separate widget. This will be harder, but will be much more maintainable in the long run. The logic will be much easier to follow, turning it off will reduce loading and runtime overhead, and future changes and improvements will be much easier to make. To do this you'd want to follow patterns that you can see in lots of other widgets e.g. gridmenu https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_gridmenu.lua#L707
It responds to calls from other widgets by this method, allowing you to dynamically change what the widget does based on calls from other widgets, settings and so on. So to do this, you'd want to add one or more API functions into top_bar following this pattern, like e.g. `includeBuildBar` which would shift sizes/placements of a couple of the features in top_bar, and nothing more, and then your external `build_bar` widget would render itself in the space you just made by doing that.

Which approach you take is your decision, and though #2 will likely take more work, it's the one I would recommend.

**Comment by roberthartmann** [comment]
I see the second version is cleaner. That is true.
I have allready made it toggable, but, eventhough I was *sure* that I did everything right do upload it... I didn't. That is why you still saw the old version without it.

Never the less it is now existing that way and I guess it still could be a good thing in a way: a POC is now existing. How to pull apart the widgets in detail is now easier than ever... BUT the metal and energy bars use so much info to calculate the real pull of resources, while being on low prio I think this calculation needs to be in a separate lua to provide the values to both top bars or the top bar needs to give the info to the BP bar.

What do you think? 

For me it feels like: just let the three bars together, they are using sooo much of the code together anyways.

**Comment by thehobojoe** [comment]
Yeah if there's a lot of code sharing it makes sense to keep it together, so long as the toggle works and it is hooked up to settings. You can use this as a reference for how to wire this up to the settings https://github.com/beyond-all-reason/Beyond-All-Reason/pull/2283/files

**Pull request closed without merging**